### PR TITLE
feat(di): plugins now have access to DI container

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

**Comment by hypeJunction** [comment]
@mrclay Lost your commit history due to conflicts, sorry.
@juho-jaakkola What do you think about this approach?

I just don't see why we need elgg() to reference an Application. It's a private class for the most part, I'd rather we turn it into a DI container.

**Comment by hypeJunction** [comment]
@jdalsem @jeabakker Neither Steve nor Juho are responsive. I'd like to get this into 3.0, it's a breaking change, but I'd rather do it before too late. I want to have services in my plugins.

**Comment by jeabakker** [comment]
> I'd like to get this into 3.0, it's a breaking change, but I'd rather do it before too late

technically breaking changes aren't allowed anymore ;)
However i think this can go in, will try to review today

**Comment by hypeJunction** [comment]
> technically breaking changes aren't allowed anymore ;)

I know, but I think it's worth it ;)

File: docs/guides/plugins.rst [hunk 1]
```diff
@@ -49,6 +49,28 @@ constants like ``__DIR__``, its return value should not change. The currently su
  * ``user_settings``
  * ``widgets``
 
+elgg-services.php
```
Comment (jeabakker) [inline docs/guides/plugins.rst]: this couldn't be done in `elgg-plugin.php`?

Comment (hypeJunction) [reply docs/guides/plugins.rst]: It could be in theory, but I like the separation that Steve introduced initially.

**Review by jeabakker — APPROVED** [review]
I don't see any objections. If @jdalsem agrees it can go in

File: docs/appendix/upgrade-notes/2.x-to-3.0.rst [hunk 2]
```diff
@@ -511,6 +511,14 @@ The following class registration related functions have been removed:
  * ``elgg_register_class``
  * ``elgg_register_classes``
 
+Dependency Injection Container
+------------------------------
+
+Plugins can now define their services and attach them to Elgg's public DI container by providing definitions in
+``elgg-services.php`` in the root of the plugin directory.
+
+``elgg()`` no longer returns an instance of Elgg application, but a DI container instance.
```
Comment (mrclay) [inline docs/appendix/upgrade-notes/2.x-to-3.0.rst]: 👍 A fine thing to break API for.

**Pull request merged**